### PR TITLE
INGM-402 Add PDO exception observers

### PR DESCRIPTION
**Pull request by pzinemanas**
### Description

This PR adds callbacks to notify when exceptions are raised in the PDO process data thread.

Fixes # INGM-402

### Type of change

- [x] Add PDO exception observers.
- [x] Add methods to pause and resume the PDO thread.

### Tests
- [x] Add new unit tests if it applies.
- [x] Run tests.

Please describe the tests that you ran to verify your changes if it applies. 

### Documentation

Please update the documentation.

- [x] Update docstrings of every function, method or class that change.
- [x] USe [type hints](https://docs.python.org/3/library/typing.html) for every function and argument.
- [x] Build documentation locally to verify changes.
- [x] Add the changes at the `[Unreleased]` section of the [CHANGELOG](CHANGELOG.md).

### Code formatting

- [x] Use black package to format the code: `black -l 100 ingeniamotion tests`. It is recommended to configure the code editor to automatically format the code using black with a max length line of 100.

### CODE REVIEW

File: ingeniamotion/pdo.py [hunk 1]
```diff
@@ -194,17 +194,43 @@ def __init__(
...
+                        e = IMException(
+                            "Stopping the PDO thread due to the following exception:"
+                            f" {e} {duration_error}"
+                        )
+                        self._notify_exceptions(e)
+                    break
                 if self._notify_receive_process_data is not None:
                     self._notify_receive_process_data()
-                time.sleep(self._refresh_rate)
+                iteration_duration = time.time() - time_start
```
Comment (MartinALL91) [inline ingeniamotion/pdo.py:230]: This is a little confusing. 
I would do:
```
  remaining_loop_time = self._refresh_rate - (time.time() - time_start)
  if remaining_loop_time > 0:
      time.sleep(remaining_loop_time)
```

Comment (pzinemanas) [reply ingeniamotion/pdo.py]: Done!

**Review by MartinALL91 — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 ### Added
 - PDOs for the EtherCAT protocol.
 - Register Poller using PDOs.
+- Callbacks to notify exceptions in the ProcessDataThread.
```
Comment (GullinBustin) [inline CHANGELOG.md:11]: Don't modify an already-released version changelog

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (pzinemanas) [reply CHANGELOG.md]: Ups... fixed!

**Review by GullinBustin — APPROVED** [review]

File: ingeniamotion/pdo.py [hunk 3]
```diff
@@ -194,17 +194,50 @@ def __init__(
...
+            try:
+                self._net.start_pdos()
+            except (ILStateError, ILError) as il_error:
+                self._pd_thread_stop_event.set()
+                im_exception = IMException(
+                    f"Could not start the PDOs due to the following exception: {il_error}"
+                )
+                self._notify_exceptions(im_exception)
+            iteration_duration: float = -1
             while not self._pd_thread_stop_event.is_set():
```
Comment (polfeliu) [inline ingeniamotion/pdo.py:210]: If there's an exception during the initialization I think the process data loop should not start

File: ingeniamotion/pdo.py [hunk 4]
```diff
@@ -194,17 +194,50 @@ def __init__(
...
+                try:
+                    self._net.send_receive_processdata()
+                except ILWrongWorkingCount as il_error:
+                    self._pd_thread_stop_event.set()
+                    self._net.stop_pdos()
+                    if iteration_duration == -1:
+                        iteration_duration = time.time() - time_start
+                    duration_error = ""
+                    if iteration_duration > self._refresh_rate:
+                        duration_error = (
```
Comment (polfeliu) [inline ingeniamotion/pdo.py:223]: Magic that should be in the doc

File: ingeniamotion/pdo.py (hunk 3 above)
Comment (pzinemanas) [reply ingeniamotion/pdo.py:210]: It will not start because the the `_pd_thread_stop_event` is set.

File: ingeniamotion/pdo.py [hunk 5]
```diff
@@ -198,7 +198,14 @@ def __init__(
 
         def run(self) -> None:
             """Start the PDO exchange"""
-            self._net.start_pdos()
+            try:
+                self._net.start_pdos()
+            except (ILStateError, ILError) as il_error:
```
Comment (MartinALL91) [inline ingeniamotion/pdo.py:203]: You could simply use ILError

**Pull request merged**